flamingo/scope.h

```diff
@@ -23,18 +23,41 @@ static flamingo_scope_t* scope_alloc(void) {
 	return scope;
 }
 
-static void scope_free(flamingo_scope_t* scope) {
-	for (size_t i = 0; i < scope->vars_size; i++) {
-		flamingo_var_t* const var = &scope->vars[i];
+static void scope_empty(flamingo_scope_t* scope) {
+	size_t const count = scope->vars_size;
+	flamingo_var_t* const vars = scope->vars;
+
+	scope->vars_size = 0;
+	scope->vars = NULL;
+
+	for (size_t i = 0; i < count; i++) {
+		flamingo_var_t* const var = &vars[i];
 
 		val_decref(var->val);
 		free(var->key);
 	}
 
-	free(scope->vars);
+	free(vars);
+}
+
+static void scope_free(flamingo_scope_t* scope) {
+	scope_empty(scope);
 	free(scope);
 }
 
+static void scope_decref(flamingo_scope_t* scope) {
+	if (scope == NULL) {
+		return;
+	}
+
+	assert(scope->ref_count > 0);
+	scope->ref_count--;
+
+	if (scope->ref_count == 0) {
+		scope_free(scope);
+	}
```

Comment:
is it possible for a scope to reach a refcount of 0 and still contain variables?